pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,495 @@
+						Err::<(), DispatchError>(Error::<T>::PendingInboundEntryNotFound.into())
+					}
+					Some(stored_inbound_entry) => {
+						let post_dispatch_entry = InboundEntry::create_post_voting_entry(
+							stored_inbound_entry,
+							expected_proof_count,
+						)?;
+
+						*storage_entry = post_dispatch_entry;
+
+						Ok(())
+					}
+				}
+			})?;
+		}
+
+		Ok(())
+	}
+
+	/// Iterates over a batch of messages and checks if the requirements for
+	/// processing each message are met.
+	pub(crate) fn process_inbound_message(
+		domain_address: DomainAddress,
+		message: T::Message,
+		router_id: T::RouterId,
+	) -> (DispatchResult, Weight) {
+		let weight = LP_DEFENSIVE_WEIGHT;
+
+		let router_ids = match Self::get_router_ids_for_domain(domain_address.domain()) {
+			Ok(r) => r,
+			Err(e) => return (Err(e), weight),
+		};
+
+		let session_id = SessionIdStore::<T>::get();
+
+		let expected_proof_count = match Self::get_expected_proof_count(&router_ids) {
+			Ok(r) => r,
+			Err(e) => return (Err(e), weight),
+		};
+
+		let mut count = 0;
+
+		for submessage in message.submessages() {
+			if let Err(e) = count.ensure_add_assign(1) {
+				return (Err(e.into()), weight.saturating_mul(count));
+			}
+
+			let message_proof = Self::get_message_proof(submessage.clone());
+
+			let inbound_entry: InboundEntry<T> = InboundEntry::create(
+				submessage,
+				session_id,
+				domain_address.clone(),
+				expected_proof_count,
+			);
+
+			if let Err(e) = inbound_entry.validate(&router_ids, &router_id.clone()) {
+				return (Err(e), weight.saturating_mul(count));
+			}
+
+			if let Err(e) = Self::upsert_pending_entry(message_proof, &router_id, inbound_entry) {
+				return (Err(e), weight.saturating_mul(count));
+			}
+
+			match Self::execute_if_requirements_are_met(
+				message_proof,
+				&router_ids,
+				session_id,
+				expected_proof_count,
+				domain_address.clone(),
+			) {
+				Err(e) => return (Err(e), weight.saturating_mul(count)),
+				Ok(_) => continue,
+			}
+		}
+
+		(Ok(()), weight.saturating_mul(count))
+	}
+
+	/// Retrieves the IDs of the routers set for a domain and queues the
+	/// message and proofs accordingly.
+	pub(crate) fn queue_outbound_message(
+		destination: Domain,
+		message: T::Message,
+	) -> DispatchResult {
+		let router_ids = Self::get_router_ids_for_domain(destination)?;
+
+		let message_proof = message.to_proof_message();
+		let mut message_opt = Some(message);
+
+		for router_id in router_ids {
+			// Ensure that we only send the actual message once, using one router.
+			// The remaining routers will send the message proof.
+			let router_msg = match message_opt.take() {
+				Some(m) => m,
+				None => message_proof.clone(),
+			};
+
+			// We are using the sender specified in the pallet config so that we can
+			// ensure that the account is funded
+			let gateway_message = GatewayMessage::<T::Message, T::RouterId>::Outbound {
+				sender: T::Sender::get(),
+				message: router_msg,
+				router_id,
+			};
```

Comment:
NIT for the next PR.

We're storing outbound messages with a `sender` field, which is a constant. We can remove the `sender` field from here, and when the message is unqueued, use it there.

Reply:
Will take care of this in the following PRs for message recovery, cheers!

Reply:
Will be done in - https://github.com/centrifuge/centrifuge-chain/pull/1970